### DISCUSSION TITLE
Guidance with WeeklyHoursLimit in the Constraint

### DISCUSSION THREAD

**Discussion by magick93** [TimefoldAI/timefold-solver#170] [General]
Hello

I'm trying (a) get some practical experience with Timefold and (b) write a constraint for the employee roster example, that will optimize based on contract min and max time, eg, work no more than n hours, no less than m hours. 

I'm reviewing a similar question on SO - https://stackoverflow.com/questions/67085636/optaplanner-custom-employee-rostering-issues-with-weeklyhourslimit-in-the-cons 

The following questions are probably very basic - I am doing a lot of reading on Optaplanner / Timefold, and really appreciate any help.

This uses a `sumDuration()` function. Is this included in Timefold?

I'm working of the Timefold nursing roster example. In this, is the `org.acme.employeescheduling.domain.EmployeeSchedule` class appropriate to use in the constraint to query for the weekly allocated work time for an employee? This model is annotated with `@PlanningSolution`. 

If this is the most appropriate model, then I'm assuming this will need to be populated from historical data, too. 
What is the name for planning that involves loading historical facts (such as the last month of shifts)? Are there any similar examples I can use to study?

Assuming I need to load historically assigned shifts, should I also add a hard constraint to prevent re-planning of historical shifts? Or is there a more native way of ignoring already assigned shifts? should I use:
-  `entitySelector` or 
- `valueSelectot` or
- `PinningFilter` or --- this looks to be the way historical shifts are filtered in `ai.timefold.solver.examples.nurserostering.domain.solver.ShiftAssignmentPinningFilter`
- `filteredSelection` ?

Thanks

**Comment by mcimbora** [comment]
Hi @magick93, welcome. We've got an employee rostering quickstart - going through the code, I think it answers most of your questions.

https://github.com/TimefoldAI/timefold-quickstarts/tree/stable/use-cases/employee-scheduling

- sumDuration: it's available in `ai.timefold.solver.core.api.score.stream.ConstraintCollectors` class
- Weekly work constraint: you can find an inspiration here `org.acme.employeescheduling.solver.EmployeeSchedulingConstraintProvider`. Maybe try to group shifts by employees & week and see if their sum is greater than the value in employee contract.
- Historic shifts: you could use concept similar to `org.acme.employeescheduling.domain.ShiftPinningFilter`, i.e. you can set a "historic" flag based on whether the input data already contains shift <-> employee assignment

**Reply by magick93** [reply]
Thank you @mcimbora - this is very helpful.

**Reply by magick93** [reply]
I have a few further questions, if that is ok.

For implementing a `SelectionFilter`, is it required that the selection type implement `java.lang.Comparable`? I notice that `ai.timefold.solver.examples.nurserostering.domain.ShiftAssignment` implements `Comparable<ShiftAssignment> ` and wonder if it due to its use in the selection filter.

**Comment by triceo** [comment]
> For implementing a `SelectionFilter`, is it required that the selection type implement `java.lang.Comparable`? I notice that `ai.timefold.solver.examples.nurserostering.domain.ShiftAssignment` implements `Comparable<ShiftAssignment> ` and wonder if it due to its use in the selection filter.

`SelectionFilter` doesn't need entities to be `Comparable`. I'd say that `ShiftAssignment` being `Comparable` is unrelated, perhaps a remnant of some earlier time; not everything in the examples strictly *needs* to be there.
